### ISSUE TITLE
Crnk does too much for us

### ISSUE THREAD

**Issue by weaselmetal** [crnk-project/crnk-framework#278]
Sorry, this is a bit lengthy.

So, you may have noticed from some of my other comments, that we created a (big) API starting back then with katharsis.
Crnk has evolved a lot since then. But many improvements and features don't work for us. In fact, we need to work around them.

I'd like to see a more layered or modular approach to features that can be turned off and on.

Some of my wishes would be:

1) Do not parse parameters that where passed in the URL. I'll use the raw ones from the request (`Map<String, Set<String>>`).
2) Do not make any assumptions or restrictions on filter parameter shape or semantics. jsonapi.org spec does make [recommendations](http://jsonapi.org/recommendations/), but they are only recommendations, and Crnk's default behavior is not happy with a filter like `GET /comments?filter[post]=1` (correct me if I'm wrong).
3) Allow me to do things, that don't comply with jsonapi spec, at least parameter-wise. We need data aggregating endpoints (building sums, averages etc) and therefore require parameters that 
allow us to 'group by' different criteria. Crnk should never drop any parameters but provide means to access non-compliant ones too.

crnk-jpa is indeed a very cool thing, if you can make do with the offered functionality.
Unfortunately, we can't. And crnk itself imposes restrictions and conventions, that make sense for crnk-jpa, but not if you have different things in mind.
I get the impression that quite some crnk-core functionality is optimized towards crnk-jpa.

Right now I'm in the situation where all parameters get parsed, some throw a parsing exception which I have to ignore, and in the end, I have to pull parameters from the raw map because it's the only way to really get everything that was passed (eg. grouping parameters).
Performance wise it may not be a big issue, but if feels terribly wrong to do so.
I am kind of writing a miniature framework inside crnk framework, to get what we need. Which is time consuming and also brittle.

Just recently I tried to add support for page[size] and page[number] parameters, instead of the default offset and limit params.
That worked fine enough, even though I had do do very, very similar things as the `OffsetLimitPagingBehavior` (shared base class maybe?).

Yet, if a URL like '/api/persons?filter[name]=earl&page[size]=10&page[number]=0' was requested, then the `next` link will be rendered as
'/api/persons?filter[name][EQ]=earl&page[size]=10&page[number]=1'. Note the added '[EQ]'. While this is fine within the crnk world, it breaks my own parameter handling.

Summing up, I have two questions:

1) Would you consider to 'do less' in crnk, and give framework users more control over which things they want to use?

2) If there are no plans to work on 1) - which behavior / classes could be changed such that I get the flexibility I need? With which classes should I start and can I make crnk use them over the default ones easily?

Thanks for reading this! Looking forward to your answers.

**Comment by remmeier** [comment]
In general I'm all for having a small crnk-core (or crnk-engine or something). To a large degree that is implemented the modular architecture, interfaces and default implementations. I think the "control" you are referring to is available. Maybe more a question of examples, documentation, customization options and having to implement that "control" vs customizing the defaults in place.

- QuerySpec should not be tied into crnk-jpa. We make use of it to integrate other services, non-relational data stores and more. But for sure it makes heavy use of it. Feel free to elaborate. 
- Having all the parameters validated and checked saves us quite a bit of code and checks and improved robustness. All too often we saw things just getting silently ignored as one example. Or having to parse values manually. 
- a more modular DefaultQuerySpec(De)Serializer would probably be a good thing to have. It is one of the primary things to customize in crnk I guess. I think to most degree it addresses the issues mentioned here and also not too much suprising coming from a QueryParams-based API. All our projects started with QuerySpec from the start and are in that regard more or less fine (with one or the other customization on our own).
- just a plain QuerySpecSerializer without validation and nothing would be not to hard if is desired. But potentially I would check whether that much flexibility in the repositories is really desired in the mid/long-term.  
- number/size next to offset/limit is something many other people could use as well I guess? Ideally that is interoperable so that the client can choose (resp. make something like crnk-jpa work with both).
- `GET /comments?filter[post]=1` is perfectly fine. 
- I think QuerySpecSerializer and Deserializer should be merged. Is on my mind for a very long time. Various paramters/customizations hold for both and should not have to be configured twice (or ignored for the serializer).
- QuerySpec serialize/deserialize should result always in the same output.
- I think it should be possible to define QuerySpecSerializer/Deserializer on a resource-level similar to how it is done for pagination. Different resources behave differently. And resources/repositories from third-party modules could come with their own implementations. But it should also be not too much of a burden to configure multiple resource.
- HttpRequestContext is fine in general I think to access request headers and non-compliant request parameters? 2.6 introduces QueryContext. Maybe there is some room in there as well to access such parameters.
- further slimming down crnk-core is currently a continuous process. getting rid of deprecated api would be one big step. a split into api/engine/defaults maybe another.
- the "group" parameter of QueryParams I still consider fairly incompatible with the JSON API specification and would rather model it as a relationship/filter or combination thereof. As such a consider HttpRequestContext  the right place for it. But something that could be discussed with JSON API community and the solution brought back to crnk as a default/example.

Some good starts (maybe in that order):
1. that num/size paging strategy and maybe a base class
2. merging QuerySpec serializer/deserializer?
3. allow per-resource behavior (not just paging, everything). Maybe something similar to RelationshipMatcher that gives flexibility to specify what is applied where (like on a java package level). 
4. introduce a FilterOperator.DEFAULT (that will not show up in the urls again like EQ)
5. make serializer/deserializer more modular/easy to customize
6. setup an advanced customization example in crnk-example
7. check where the parsing/validation can improve (parsing "null" did come up lately)
8. get rid of deprecated things
8. modularize/split crnk-core a bit more.

on my side I currently work on 2.6 in terms of performance and supported frameworks (like ratpack/vertx) to widen the audience a bit. planning & contributions are alway very welcomed. most things are also rather incremental than major upgrades. a crnk 3.0 would manifest eventually I think to get rid of deprecations and further moduarlize crnk-core.

**Comment by weaselmetal** [comment]
Thanks for the reply @remmeier!

Gotta read your reply again and think things through. One thing though: wouldn't
`GET /comments?filter[post]=1`
make the parameter parsing try to construct a `post` resource from the value '1' - as post is a related object on the comment?

**Comment by remmeier** [comment]
yes, if post is a relation, then currently it would be filter[post.id]=1. One of the things that with a more modular/customizable QuerySpecSerializer could be easily addressed.

**Comment by weaselmetal** [comment]
right. `QuerySpecSerializer` is very under the hood currently.
I don't see any way to use a custom `QuerySpecSerializer`.
`JsonApiUrlBuilder` always creates a `DefaultQuerySpecSerializer`.

**Comment by remmeier** [comment]
yes, it has also grown a bit and does not look overly pretty. so far in the projects I'm involved with we more or less stick to the defaults. But something that can be cleaned up fairly quickly I think.

**Comment by weaselmetal** [comment]
Time for the big answer to your initial response.

> - QuerySpec should not be tied into crnk-jpa. We make use of it to integrate other services, non-relational data stores and more. But for sure it makes heavy use of it. Feel free to elaborate.
> - Having all the parameters validated and checked saves us quite a bit of code and checks and improved robustness. All too often we saw things just getting silently ignored as one example. Or having to parse values manually.

QuerySpec tries to infer the target type for filter parameters from the model (the resource) that is being retrieved. That works well within crnk-jpa. It doesn't work well, when filter parameters don't match model attributes or relationships. Example: `filter[person][hasChildren]=true` where `hasChildren` is not an attribute on the resource, but should impose appropriate DB-filter criteria.
In our API, we use a completely different approach. We have a `FilterDefinition` object per endpoint, which defines optional and required parameters, their types, and their default value (if any).
In theory this `FilterDefinition` would also allow us to auto generate a documentation for all accepted filter parameters. But this approach does not align well with QuerySpec.

> - a more modular DefaultQuerySpec(De)Serializer would probably be a good thing to have. It is one of the primary things to customize in crnk I guess.

Agreed. Either the (De)Serializer is highly configurable, or crnk offers different implementations to choose from. 

> - just a plain QuerySpecSerializer without validation and nothing would be not to hard if is desired.

For our API I now wrote a Deserializer that does nothing but collect the passed params and organize it into filter, page, include, ... parameters. All parameter values are kept as String by intention. I think that such a Deserializer should be available by default within crnk.
In any case, I think that QuerySpec should support access to all passed parameters.
If I pass something like `GET api/person?logResponsePayload=true` I'd like to be able to retrieve that parameter from the QuerySpec object.

> - I think QuerySpecSerializer and Deserializer should be merged. Is on my mind for a very long time. Various paramters/customizations hold for both and should not have to be configured twice (or ignored for the serializer).

That would make a lot of sense I think. Even more so since it's currently impossible to register a custom QuerySpecSerializer.

> - I think it should be possible to define QuerySpecSerializer/Deserializer on a resource-level similar to how it is done for pagination. Different resources behave differently. And resources/repositories from third-party modules could come with their own implementations. But it should also be not too much of a burden to configure multiple resource.

I partly agree. I'd want to specify an API-wide default behavior and override that where needed.
In general I tend to say that pagination and parameter handling should be the same for the entire API.
But as we know, from time to time you just have to forgo consistency.

> - HttpRequestContext is fine in general I think to access request headers and non-compliant request parameters? 2.6 introduces QueryContext. Maybe there is some room in there as well to access such parameters.

I don't agree. In our case we would have to read parameters from QuerySpec and the raw request many times.
Of course, for non-standard parameters crnk has no way to figure out the target type (which it shouldn't / can't do anyway for some parameters, see my first paragraph in this comment).
I think crnk should be a framework that allows you create an API that complies with jsonapi.org spec. It should not enforce compliance with that specification. It would surely increase the flexibility of app development. (Also this discussion is about data-grouping, but there's no conclusion to it: http://discuss.jsonapi.org/t/representing-non-resourceful-aggregated-data/431)

> - further slimming down crnk-core is currently a continuous process. getting rid of deprecated api would be one big step. a split into api/engine/defaults maybe another.

That's a great plan to follow up upon. The removal of deprecated classes is worrying for us, though. It's quite hard (well, impossible describes it better) to move endpoints that have been using QueryParams over to use QuerySpec, if those endpoints use parameters that ceased to exist in QuerySpec, like those `group` parameters. I have started to pull **all** parameters from the raw request for both old and new style endpoints. Because there is nothing else I can do without breaking the API for clients. But this refactoring is very time consuming.

In general I'm happy to help where I can. But tasks need to be well defined, well isolated and accomplishable in a reasonable amount of time, ie be small enough.
And, last but not least, our contributions would ideally reduce the amount of code we have written in our API to achieve a behavior that works for us.

**Comment by remmeier** [comment]
> Example: filter[person][hasChildren]=true where hasChildren is not an 
> attribute on the resource, but should impose appropriate DB-filter criteria.

There are two things here:
- it should rather be filter[person][children][exists]=true or just filter[children][exists]=true where the "has" is implemented as operator next to possible other operators. This way the entire thing stays within the known set of fields. I think it is actually quite rare for a filter parameter not be associated with one of the fields in one way or the other (next to general full-text search where JSON API may should rather introduce another parameter like "search").
- there is one piece missing so far. Wanted to do that for a while now. The field typically gives the type of the parameter. But it should be the applied operator that has the final say about which type is expected as parameter. 
Those things together cover quite a wide range of what I have seend the past year. Or do you have many other use cases that would not fit the pattern?

> We have a FilterDefinition object per endpoint

@JsonApiField allows to specify what is sortable, filterable. That information is available from crnk-meta as well to clients. No support yet to query the available operators tough. We make actually heavy use of that in a bit of a more advanced flavor of the crnk-ui. 

not sure about FilterDefinition. A small next step in the current architecture would be a ResourceFieldType.FILTER do specify the few(?) things that not match attributes/relationships. There might be room for more in terms of a FilterDefinition. To some degree I would also like to see support for nested AND/OR/NOT with filters/operators/value. QuerySpec is already prepared for that, just no support in the (de)serializer.

> For our API I now wrote a Deserializer that does nothing but collect

could be a configuration flag or a customization of the existing one then (after the merging)

> QuerySpec should support access to all passed parameters.

a bit reluctant in this area (with grouping clearly been one of the harder ones in a gray area). I have seen a far share of abuse and not necessarly want to make it too easy to go down this path. In many cases meta-information, HTTP header, server flags, or just JSON API-conform modelling seemed to be the better fit.  But a bit of a cleanup of that HttpRequestContext handling is on its way with the reactive part (since there is no thread-local anymore), so things have to slightly change anyway. And would not mind a PR to add all unknown parameters to pagin links. That seems reasonable.

> move endpoints that have been using QueryParams 

QueryParams is still here to stay for quite a while. It also does not cause too much trouble since QuerySpec is only one possible implementations. QueryAdapter is the main contract within the crnk engine. But other things like the repository annotations really have to go away as they cause issues within the engine part.

> In general I'm happy to help where I can. But tasks need to be well defined, well isolated

page[number]/page[size] could be one thing. Ideally compatible with offset/limit, meaning the possibility to maybe cast it to a OffsetLimtPagingSpec (and the other way around). This way something like crnk-jpa can work with both and internally can continue to use of offset/limit. At some point in the future there may also be a jpa specfic behavior to support other paging behavior. 

another thing could be that paging/unknown parameter handling.

I will take a look at the FilterOperator and maybe merging the serializers.

Then we go from there. For the individual parts we can create dedicated tickets and dicuss in more detail where necessary to not overload this one.

**Comment by weaselmetal** [comment]
Well, `filter[person][hasChildren]=true` was an overly simplified example.
A better example would be `filter[person][hasSilverSpoons]=true`.

Which, according to your reasoning should be something like
`filter[person][apartment][inventory][spoons][material]=silver`. 

We do have cases where we need to filter on data that is 4 hops (and more) away from the root object.
While the latter approach is certainly not wrong, it was our design decision to rather use `filter[person][hasSilverSpoons]=true`. **We even have filters that aren't expressible through a relationship chain at all!** This is where your suggested approach ultimately stops working. Not every relationship should / can be modeled bidirectionally. Which shouldn't prevent one from defining a filter (where you will have to start with a DB query on a totally different table).
Our API is not only used by our own app, we also have customers that communicate with our API directly. So a simplification of otherwise complex filter criteria is definitely important for us. Therefore

> I think it is actually quite rare for a filter parameter not be associated with one of the fields in one way or the other.

... is an assumption which doesn't hold true. We have plenty such parameters. 

And didn't mean to say that crnk should also start using `FilterDefinition` objects (although I am very happy with that idea), I was just trying to explain why the type-from-resource-deduction doesn't work well for us and how we infer the type.

I'm happy to hear you're open for change!

**Comment by remmeier** [comment]
QuerySpec serialize/deserializer have been merged to have everything at one place, which in turn allows to more easily implement more customized url/parameter handling as requested here. And paging can be configured globally. There are a few smaller follow-ups:

https://github.com/crnk-project/crnk-framework/issues/307
https://github.com/crnk-project/crnk-framework/issues/306
https://github.com/crnk-project/crnk-framework/issues/291

Probably worthwhile to close this in favor of smaller more targeted tickets.
